src/background/http-instrument.ts

```diff
@@ -1,687 +1,693 @@
   }
-  return stacktrace.join("\n");
-}
 
-const httpRequestHandler = function(reqEvent, crawlID) {
-  const httpChannel = reqEvent.subject.QueryInterface(Ci.nsIHttpChannel);
-
-  // Save HTTP redirect events. Requires FF 49+
-  // Events are saved to the `http_redirects` table, and map the old
-  // request/response channel id to the new request/response channel id.
-  // Implementation based on: https://stackoverflow.com/a/11240627
-  const oldNotifications = httpChannel.notificationCallbacks;
-  let oldEventSink = null;
-  httpChannel.notificationCallbacks = {
-    QueryInterface: XPCOMUtils.generateQI([
-      Ci.nsIInterfaceRequestor,
-      Ci.nsIChannelEventSink,
-    ]),
-
-    getInterface(iid) {
-      // We are only interested in nsIChannelEventSink,
-      // return the old callbacks for any other interface requests.
-      if (iid.equals(Ci.nsIChannelEventSink)) {
-        try {
-          oldEventSink = oldNotifications.QueryInterface(iid);
-        } catch (anError) {
-          loggingDB.logError(
-            "Error during call to custom notificationCallbacks::getInterface." +
-              JSON.stringify(anError),
-          );
-        }
-        return this;
+  public run(crawlID, saveJavascript, saveAllContent) {
+    const allTypes: ResourceType[] = [
+      "beacon",
+      "csp_report",
+      "font",
+      "image",
+      "imageset",
+      "main_frame",
+      "media",
+      "object",
+      "object_subrequest",
+      "ping",
+      "script",
+      // "speculative",
+      "stylesheet",
+      "sub_frame",
+      "web_manifest",
+      "websocket",
+      "xbl",
+      "xml_dtd",
+      "xmlhttprequest",
+      "xslt",
+      "other",
+    ];
+
+    const requestStemsFromExtension = details => {
+      return (
+        details.originUrl && details.originUrl.indexOf("moz-extension://") > -1
+      );
+    };
+
+    /*
+     * Attach handlers to event listeners
+     */
+
+    this.onBeforeRequestListener = details => {
+      const blockingResponseThatDoesNothing: BlockingResponse = {};
+      // Ignore requests made by extensions
+      if (requestStemsFromExtension(details)) {
+        return blockingResponseThatDoesNothing;
       }
+      const pendingRequest = this.getPendingRequest(details.requestId);
+      pendingRequest.resolveBeforeRequestEventDetails(details);
+      const pendingResponse = this.getPendingResponse(details.requestId);
+      pendingResponse.resolveBeforeRequestEventDetails(details);
+      if (saveAllContent) {
+        pendingResponse.addResponseResponseBodyListener(details);
+      } else if (
+        saveJavascript &&
+        this.mayYieldAJSResponse(details.type, details.url)
+      ) {
+        pendingResponse.addResponseResponseBodyListener(details);
+      }
+      return blockingResponseThatDoesNothing;
+    };
+    browser.webRequest.onBeforeRequest.addListener(
+      this.onBeforeRequestListener,
+      { urls: ["http://*/*", "https://*/*"], types: allTypes },
```

Comment:
We can use `<all_urls>` to match everything. I opted for the http(s) pattern because of the name of the file :) - maybe call it `request-instrument`?

Reply:
Now using `<all_urls>`

Reply:
Good call! It looks like that will also capture `data` urls, which I suspect are misused in various ways. `request-instrument` is fine; the current name goes back to FourthParty (https://github.com/fourthparty/fourthparty/blob/master/extension/lib/http-instrument.js).

Reply:
Let's think about the renaming a bit more before acting on it. `request-instrument` implies a `request-instrument`, and even the official docs about the webRequest API states `Add event listeners for the various stages of making an HTTP request` even though technically other protocols monitored, it may all be HTTP under the hood, so `http-instrument` is good enough for now :)